Migrate away from deprecated kustomize features

james-callahan opened the pull request:
### Issue

<!-- Please link the GitHub issues related to this PR, if available -->

https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/3090

### Description

<!--
Please explain the changes you made here.

Help your reviewers my guiding them through your key changes,
implementation decisions etc.
You can even include snippets of output or screenshots.

A good, clear description == a faster review :)
-->

This PR cleans up various things in the kustomize manifests. See individual commits for descriptions of the modifications.

In particular:
  - we now don't have any deprecation warnings when you run `kustomize build config/default` with kustomize v5.0.0.
  - there's now only one place you need to go to enable/disable the webhook/cert-manager integration

### Checklist
- [ ] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `docs` directory)
- [x] Manually tested
- [ ] Made sure the title of the PR is a good description that can go into the release notes

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [x] Refactored something and made the world a better place :star2:

james-callahan committed:
Clean up kustomizations

  - Always include apiVersion+kind of kustomization files
  - `patchesStrategicMerge` was deprecated, use `patches` instead
  - `bases` was deprecated, use `resources` instead
  - Order `configurations` first

james-callahan committed:
Use commonLabels

james-callahan committed:
Add ports declaration to controller container

james-callahan committed:
Consolidate cert-manager kustomize pieces into the one component

james-callahan committed:
Consolidate webhook kustomize pieces into the one component

james-callahan committed:
Kustomize deprecated `vars`, use `replacements`

james-callahan committed:
Use CRD v1 form

kustomize also requires that we specify a namespace here, or else it won't add one

james-callahan committed:
caBundle doesn't need to be set

And it causes the resource to show up in e.g. ArgoCD as Out-of-Sync

james-callahan committed:
Remove hardcoded namespaces

kustomize will fill these in for us

james-callahan commented:
Rebased on v2.5.0.

james-callahan commented:
Rebased on top of main post-v2.6.1

james-callahan commented:
@kishorj  / @M00nF1sh could you have a look at this?

drGrove commented:
This will solve warnings in our CI system. Can someone take a look?

drGrove commented:
/remove-lifecycle rotten

james-callahan commented:
rebased on v2.7.1

james-callahan commented:
> PR needs rebase.

And rebased again on main

drGrove commented:
/remove-lifecycle-rotten

drGrove commented:
/reopen

james-callahan commented:
/reopen

archer-65 commented:
/reopen

james-callahan commented:
/reopen

james-callahan commented:
/reopen

james-callahan commented:
rebased on top of v2.9.0

james-callahan commented:
/reopen

Can someone please look at this? It's been waiting for over a year and a half. @shraddhabang @zac-nixon ?

james-callahan commented:
rebased on v2.11.0.....

james-callahan commented:
Rebased on v2.12.0

zac-nixon commented:
Hello, I apologize that we didn't attend to this PR sooner. I have some concerns that the proposed changes are not backwards compatible. Does kustomize define replacements for the functionality you're removing?

james-callahan commented:
> Does kustomize define replacements for the functionality you're removing?

Which functionality is being removed?

james-callahan commented:
Rebased on v2.13.2...

james-callahan commented:
Rebased on v2.13.3

zac-nixon reviewed (commented):
Sorry for not elaborating on these issues in my previous comment.

james-callahan commented on config/certmanager/certificate.yaml:
```diff
@@ -14,10 +14,10 @@ kind: Certificate
   dnsNames:
-    - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc
-    - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc.cluster.local
+    - webhook-service.default.svc
```
The `default` is swapped out by the `replacements` block here: https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/3091/files#diff-c006b1ef995eeeb57408cce42bd84db8dfc0cbe18df03222857665cebb1ac6fdR31

james-callahan commented on config/certmanager/cainjection_in_targetgroupbindings_patch.yaml:
```diff
@@ -4,5 +4,6 @@ apiVersion: apiextensions.k8s.io/v1
   annotations:
-    cert-manager.io/inject-ca-from: $(CERTIFICATE_NAMESPACE)/$(CERTIFICATE_NAME)
+    # `default` and `serving-cert` may be substituted by kustomize
+    cert-manager.io/inject-ca-from: default/serving-cert
```
The `default` is swapped out by the `replacements` block here: https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/3091/files#diff-c006b1ef995eeeb57408cce42bd84db8dfc0cbe18df03222857665cebb1ac6fdR31

james-callahan commented on config/webhook/manifests.yaml:
```diff
@@ -9,7 +9,6 @@ webhooks:
     clientConfig:
       service:
         name: webhook-service
-        namespace: system
```
Because it's not in the `system` namespace (and it's injected automatically by kustomize to the correct namespace as long as it's not set)

james-callahan commented:
/remove-lifecycle stale

james-callahan committed:
clean up AGA CRD handling

  - the CRD itself was duplicated
  - the makefile rule copied the file twice and then moved it to helm directory; do it in one shot

james-callahan committed:
clean up gateway CRD handling

  - the CRD itself was duplicated
  - the makefile rule copied the file twice and then moved it to helm directory; do it in one shot
  - update docs to use kustomize invocation
